ocaml/typing/jkind.ml

```diff
@@ -2295,32 +2315,66 @@ module Violation = struct
 
   let report_with_name ~name = report_general "" pp_print_string name
 end
+
 (******************************)
 (* relations *)
 
-let rec equate_or_equal ~allow_mutation (t : t) (t' : t) =
-  match t, t' with
-  | Type ty, Type ty' -> Type.equate_or_equal ~allow_mutation ty ty'
-  | ( Arrow { args = args1; result = result1 },
-      Arrow { args = args2; result = result2 } ) ->
-    equate_or_equal ~allow_mutation result1 result2
-    && List.for_all2 (equate_or_equal ~allow_mutation) args1 args2
-  | Type _, Arrow _ | Arrow _, Type _ -> false
+let equate_or_equal ~allow_mutation (t : t) (t' : t) =
+  let rec go k k' =
+    match k, k' with
+    | Type ty, Type ty' ->
+      Type.Jkind_desc.equate_or_equal ~allow_mutation ty ty'
+    | ( Arrow { args = args1; result = result1 },
+        Arrow { args = args2; result = result2 } ) ->
+      let arg_eqs = List.map2 go args1 args2 in
```

Comment:
Why build a list of bools with `map2` when you could build a single bool using `for_all2`?